lxd/network/driver_ovn.go

```diff
@@ -1430,6 +1430,52 @@ func (n *ovn) startUplinkPort() error {
 	return fmt.Errorf("Failed starting uplink port, network type %q unsupported as OVN uplink", uplinkNet.Type())
 }
 
+// configurePortsForVLAN configures the necessary ports to handle VLAN traffic between
+// the uplink network and OVN.
+func (n *ovn) configurePortsForVLAN(vlan string, update bool) error {
+	if vlan == "" || update && n.config["vlan"] == vlan {
+		return nil
+	}
+
+	uplinkNet, err := LoadByName(n.state, api.ProjectDefaultName, n.config["network"])
+	if err != nil {
+		return err
+	}
+
+	ovs := openvswitch.NewOVS()
+	vars := n.uplinkPortBridgeVars(uplinkNet)
+	uplinkHostName := GetHostDevice(uplinkNet.Config()["parent"], uplinkNet.Config()["vlan"])
```

Comment:
Hrm, so referring to my question https://github.com/canonical/lxd/issues/12234#issuecomment-2379610382 that we didn't discuss yet:

> Should we add a vlan setting to the uplink network or each OVN network?

You've gone with adding the `vlan` setting to each OVN network in this PR.

Because each uplink network can be used by multiple OVN networks, this means there is the possibility of using multiple VLANs on a single uplink (which in principle would work as we are trunking VLANs through the uplink OVS bridge into the uplink interface).

But I don't think using the uplink's vlan setting in the call to ` GetHostDevice` makes sense in that case.

I also see some other places in the existing code where we've called `uplinkHostName := GetHostDevice` and used the uplink's VLAN setting which needs to be reviewed as part of this PR.

Right now, say I had an eth0.101 interface (a physical port eth0, with an untagged VLAN 101 interface layered ontop) which I want to use for my uplink, you could create a physical network that has `parent=eth0` and `vlan=101` and when passing directly to an instance, LXD would check if the eth0.101 interface existed, and if not create it, before passing it to the instance.

For OVN uplink networks, I can't see that LXD would create the eth0.101 interface if it doesn't exist, but if it does exist we would then connect eth0.101 to the intermediate uplink OVS bridge like any other port.

But in this PR you are introducing per-OVN network uplink VLAN settings, so this VLAN setting on the uplink network conceptually conflicts with what we are trying to do, as you are trying to allow multiple OVN networks to trunk multiple VLANs through a single uplink network, whilst at the same time allowing the uplink port to be configured with a single untagged VLAN (which wouldn't work with tagged frames).

But stepping back a bit further, lets forget the uplink's existing `vlan` setting for a moment.

Say we go ahead with your proposal here of adding `vlan` to the OVN network itself, then if I have two OVN networks both connected to the same uplink but using two different uplink VLAN IDs, how do we allocate an IP for the OVN network's external (volatile) IP on the uplink network, as each OVN network is effectively connected to different uplink networks (because each VLAN on the uplink network is likely going to have different IP ranges, or at least conceptually separate IP ranges).

So going back to my question above:

> Should we add a vlan setting to the uplink network or each OVN network?

I actually think we should keep the `vlan` setting on the uplink network. 
Don't introduce a `vlan` setting on the OVN network.

This would then require that the administrator define separate uplink networks in LXD for each VLAN on the parent interface, along with the separate IP range settings (which is what we want).

So the question then becomes: How should we use the `vlan` setting on the uplink network?

Also it seems that whilst the OVN network doesn't create the vlan interface (e.g. eth0.101), if the uplink is a physical network it does seem to create it:

https://github.com/canonical/lxd/blob/main/lxd/network/driver_physical.go#L318

So, because we already have `uplinkHostName := GetHostDevice(uplinkConfig["parent"], uplinkConfig["vlan"])` in `startUplinkPortPhysical` could it be that if you use an OVN network with a physical network as an uplink using a dedicated interface (not a bridge) that VLAN support already works?

Reply:
>  could it be that if you use an OVN network with a physical network as an uplink using a dedicated interface (not a bridge) that VLAN support already works?

I've just tested this and it works already, but only for physical uplinks, not bridged uplinks.

We should probably just add the last step from https://github.com/canonical/lxd/issues/12234#issuecomment-2379604497 `startUplinkPortBridgeNative` and the equivalent to `startUplinkPortBridgeOVS` to this PR instead:

> bridge vlan add dev <interface from the uplink OVS bridge to uplink> vid 101

Reply:
Thanks @tomponline. I've pushed changes which better handle the case where the uplink parent is a bridge and the `vlan` config is set -- we now avoid using the physical interface lxd creates on top of the bridge. This is WIP as I have a couple questions:

> We should probably just add the last step from https://github.com/canonical/lxd/issues/12234#issuecomment-2379604497 startUplinkPortBridgeNative and the equivalent to startUplinkPortBridgeOVS to this PR instead:

Can you help me to understand what we would need to add to `startUplinkPortBridgeOVS`?

Also, since the `vlan` config is set on the uplink network, I'm not sure what the best approach is to ensure ports are reconfigured when the uplink network `vlan` config is changed. Would we need to add a check in the network update handler which notifies cluster members to update their ports if (1) the updated network is used as an uplink for an OVN network and (2) the `vlan` config is changed? This seems a bit indirect so I'm wondering if there is something better we can do.